backend/sftp/file.go

```diff
@@ -0,0 +1,341 @@
+func (f *File) Location() vfs.Location {
+	return &Location{
+		fileSystem: f.fileSystem,
+		path:       path.Dir(f.path),
+		Authority:  f.Authority,
+	}
+}
+
+// Move/Copy Operations
+
+// MoveToFile puts the contents of File into the targetFile passed using File.CopyToFile.
+// If the copy succeeds, the source file is deleted. Any errors from the copy or delete are
+// returned.
+// If the given location is also sftp AND for the same user and host, the sftp Rename method is used, otherwise
+// we'll do a an io.Copy to the destination file then delete source file.
+func (f *File) MoveToFile(t vfs.File) error {
+	// sftp rename if vfs is sftp and for the same user/host
+	if f.fileSystem.Scheme() == t.Location().FileSystem().Scheme() &&
+		f.Authority.User == t.(*File).Authority.User &&
+		f.Authority.Host == t.(*File).Authority.Host {
+		//ensure destination exists before moving
+		exists, err := t.Location().Exists()
+		if err != nil {
+			return err
+		}
+		if !exists {
+			//it doesn't matter which client we use since they are effectively the same
+			client, err := f.fileSystem.Client(f.Authority)
+			if err != nil {
+				return err
+			}
+			err = client.MkdirAll(t.Location().Path())
+			if err != nil {
+				return err
+			}
+		}
+		return f.sftpRename(t.(*File))
+	}
+
+	//otherwise do copy-delete
+	if err := f.CopyToFile(t); err != nil {
+		return err
+	}
+	return f.Delete()
+}
+
+// MoveToLocation works by creating a new file on the target location then calling MoveToFile() on it.
+func (f *File) MoveToLocation(location vfs.Location) (vfs.File, error) {
+
+	newFile, err := location.FileSystem().NewFile(location.Volume(), path.Join(location.Path(), f.Name()))
+	if err != nil {
+		return nil, err
+	}
+
+	err = f.MoveToFile(newFile)
+	if err != nil {
+		return nil, err
+	}
+	return newFile, nil
+}
+
+// CopyToFile puts the contents of File into the targetFile passed.
+func (f *File) CopyToFile(file vfs.File) error {
+
+	if err := utils.TouchCopy(file, f); err != nil {
+		return err
+	}
+	//Close target to flush and ensure that cursor isn't at the end of the file when the caller reopens for read
+	if cerr := file.Close(); cerr != nil {
+		return cerr
+	}
+	//Close file (f) reader
+	return f.Close()
+}
+
+// CopyToLocation creates a copy of *File, using the file's current path as the new file's
+// path at the given location.
+func (f *File) CopyToLocation(location vfs.Location) (vfs.File, error) {
+
+	newFile, err := location.FileSystem().NewFile(location.Volume(), path.Join(location.Path(), f.Name()))
+	if err != nil {
+		return nil, err
+	}
+
+	if err := utils.TouchCopy(newFile, f); err != nil {
+		return nil, err
+	}
+	//Close target file to flush and ensure that cursor isn't at the end of the file when the caller reopens for read
+	if cerr := newFile.Close(); cerr != nil {
+		return nil, cerr
+	}
+	//Close file (f) reader
+	if cerr := f.Close(); cerr != nil {
```

Comment:
Same here, this is making decisions for the API user around the state of the source file. Maybe this is just keeping things consistent with the other implementations, but we should probably clarify the behavior in docs.

Reply:
Same response here.